### PR TITLE
[Nested Tensor]Nested reshape

### PR DESCRIPTION
**Pull request by drisspg**
## Summary
This is a copy of this PR: #82754 
In order to continue working on this and landing. No changes have yet been made

---
Previous reshape [#80981](https://github.com/pytorch/pytorch/pull/80981) is ok for forward, but needs improvement for backward: need to handle "sometimes view sometimes copy" behavior.

This pull request fixes it by:
1. add a new alias dispatch key `CompositeImplicitAutogradNestedTensor`, which ideally would work as nested-tensor version of `CompositeImplicitAutograd`
2. register `reshape_nested` to `reshape` by `CompositeImplicitAutogradNestedTensor`

Side changes:
* add contiguous memory format support to `clone_nested`
* add `view_nested`
* add `reshape_as_nested`

Fix issue [#83041](https://github.com/pytorch/pytorch/issues/83041)

### CODE REVIEW

File: c10/core/DispatchKeySet.cpp [hunk 1]
```diff
@@ -46,14 +46,13 @@ bool isBackendDispatchKey(DispatchKey t) {
...
 // maps to [math_dispatch_keyset x full_backend_mask]
-constexpr DispatchKeySet math_dispatch_keyset = backend_dispatch_keyset |
-    autograd_dispatch_keyset |
-    // See Note [NestedTensor Not Included in Backend Keys]
-    // The caveat to that note is that nested_tensor is a special case
-    // where we would like to support composite implicit kernels but not
-    // explicit kernels therefore we manually add the key to the
-    // math_dispatch_keyset
-    DispatchKeySet{DispatchKey::NestedTensor};
+constexpr DispatchKeySet math_dispatch_keyset =
```
Comment (drisspg) [inline c10/core/DispatchKeySet.cpp:49]: @bdhirsh Copying over to maintain provenance

Comment (drisspg) [reply c10/core/DispatchKeySet.cpp:49]: > Did this remove the ability for nested tensors to run in "regular" Composite kernels. We should have both right? And just have CompositeImplicitAutogradNestedTensor take precedence. @bdhirsh

Comment (drisspg) [reply c10/core/DispatchKeySet.cpp:49]: > I think yes, but I have modified codegen in such a way that CompositeImplicitAutogradNetsedTensor kernel got registered even if only CompositeImplicitAutograd is specified. Take `softmax` as an example, in native_functions.yaml it has:
> 
> ```
> - func: softmax.int(Tensor self, int dim, ScalarType? dtype=None) -> Tensor
>   variants: function, method
> ```
> 
> which makes it CompositeImplicitAutograd. A `torch._C._dispatch_dump("aten::softmax.int")` call gives us:
> 
> ```
> name: aten::softmax.int
> schema: aten::softmax.int(Tensor self, int dim, int? dtype=None) -> Tensor
> debug: registered at aten/src/ATen/RegisterSchema.cpp:6
> alias analysis kind: FROM_SCHEMA
> Functionalize: registered at aten/src/ATen/RegisterFunctionalization_2.cpp:20485 :: (Tensor _0, int _1, int? _2) -> Tensor _0 [ boxed unboxed ]
> Named: registered at ../aten/src/ATen/core/NamedRegistrations.cpp:11 :: (none) [ fallthrough boxed ]
> Tracer: registered at ../torch/csrc/autograd/generated/TraceType_3.cpp:14597 :: (Tensor _0, int _1, int? _2) -> Tensor _0 [ boxed unboxed ]
> AutocastCUDA: registered at ../aten/src/ATen/autocast_mode.cpp:328 :: (Tensor _0, int _1, int? _2) -> Tensor _0 [ boxed unboxed ]
> CompositeImplicitAutograd[alias]: registered at aten/src/ATen/RegisterCompositeImplicitAutograd.cpp:6983 :: (Tensor _0, int _1, int? _2) -> Tensor _0 [ boxed unboxed ]
> CompositeImplicitAutogradNestedTensor[alias]: registered at aten/src/ATen/RegisterCompositeImplicitAutogradNestedTensor.cpp:6922 :: (Tensor _0, int _1, int? _2) -> Tensor _0 [ boxed unboxed ]
> ```
> 
> Which has CompositeImplicitAutogradNestedTensor registered as well

Comment (drisspg) [reply c10/core/DispatchKeySet.cpp:49]: > hmm I'm not sure that this is what we want.
> 
> I think of `CompositeImplicitAutograd` as "all backends + layout keys, plus all autograd keys".
> 
> And then now, I think of `CompositeImplicitAutogradNestedTensor` as "all nested backends, plus the nested autograd key".
> 
> So in that view, `CompositeImplicitAutograd` should be a strict superset of `CompositeImplicitAutogradNestedTensor`: if a kernel is registered directly to `CompositeImplicitAutograd`, that should still cause all of the nested tensor dispatch keys to get that kernel too.
> 
> We also have a handful of native_function entries that explicit use `CompositeImplicitAutograd`, so this would cause NestedTensor to lose those kernels. e.g.
> 
> https://github.com/pytorch/pytorch/blob/1f38225b56b873c944196241ea448445a61798fd/aten/src/ATen/native/native_functions.yaml#L3951
> 
> .

File: aten/src/ATen/native/native_functions.yaml [hunk 2]
```diff
@@ -6943,6 +6939,7 @@
   dispatch:
     ZeroTensor, CPU, CUDA, Meta, QuantizedCPU, QuantizedCUDA, MPS: view
     MkldnnCPU: mkldnn_view
+    NestedTensorCPU, NestedTensorCUDA: view_nested
```
Comment (drisspg) [inline aten/src/ATen/native/native_functions.yaml:6942]: > hmm I'm not sure that this is what we want.
> 
> I think of `CompositeImplicitAutograd` as "all backends + layout keys, plus all autograd keys".
> 
> And then now, I think of `CompositeImplicitAutogradNestedTensor` as "all nested backends, plus the nested autograd key".
> 
> So in that view, `CompositeImplicitAutograd` should be a strict superset of `CompositeImplicitAutogradNestedTensor`: if a kernel is registered directly to `CompositeImplicitAutograd`, that should still cause all of the nested tensor dispatch keys to get that kernel too.
> 
> We also have a handful of native_function entries that explicit use `CompositeImplicitAutograd`, so this would cause NestedTensor to lose those kernels. e.g.
> 
> https://github.com/pytorch/pytorch/blob/1f38225b56b873c944196241ea448445a61798fd/aten/src/ATen/native/native_functions.yaml#L3951
> 
> .

File: c10/core/DispatchKey.h [hunk 3]
```diff
@@ -439,6 +439,8 @@ enum class DispatchKey : uint16_t {
   Autograd,
   CompositeImplicitAutograd, // registered at
   // build/aten/src/ATen/RegisterCompositeImplicitAutograd.cpp
+  CompositeImplicitAutogradNestedTensor, // registered at
```
Comment (drisspg) [inline c10/core/DispatchKey.h:442]: > hmmm I think you'll need to make internal fbcode changes, since this adds a new codegen'd file.
> 
> I would fbgs search for `RegisterCompositeImplicitAutograd.cpp` to see where that file gets hardcoded into the build system, to see where you need to add it

File: torchgen/gen.py [hunk 4]
```diff
@@ -1220,7 +1225,10 @@ def compute_registration_declarations(
         # TODO: What exactly is the semantics of the 'dispatch' field?
         "dispatch": str(
             {k for k, v in backend_indices.items() if v.has_kernel(f)}
-            != {DispatchKey.CompositeImplicitAutograd}
+            != {
```
Comment (drisspg) [inline torchgen/gen.py:1228]: > This looks like it'll change the output of "dispatch" for existing native functions that only have a `CompositeImplicitAutograd` entry.
> 
> Feels kinda ugly, but I guess you want to check for `... != {CompositeImplicitAutograd} and ... != {CompositeImplicitAutograd, CompositeImplicitAUtogradNestedTensor}`

File: torchgen/model.py [hunk 5]
```diff
@@ -656,8 +660,9 @@ def from_yaml(
                     )
                     if (
                         dispatch_key is DispatchKey.CompositeImplicitAutograd
-                        and v == cpp.name(func)
-                    ):
+                        or dispatch_key
+                        is DispatchKey.CompositeImplicitAutogradNestedTensor
```
Comment (drisspg) [inline torchgen/model.py:664]: > I think we can kill this if you follow feedback from https://github.com/pytorch/pytorch/pull/82754/files#r950515109

File: torchgen/model.py [hunk 6]
```diff
@@ -902,6 +925,7 @@ def __post_init__(self) -> None:
     def has_composite_kernel(self) -> bool:
         return (
             self.has_composite_implicit_autograd_kernel
+            or self.has_composite_implicit_autograd_nested_tensor_kernel
```
Comment (drisspg) [inline torchgen/model.py:928]: > Is it the case that "if an op has a CompositeImplicitAutogradNestedTensor kernel, then it always has a CompositeImplicitAutograd kernel"?
> 
> If it is, then we can probably take advantage of that to simplify this: just assert that in `model.py`, and you don't need this extra check.

File: torchgen/model.py [hunk 7]
```diff
@@ -2559,6 +2586,14 @@ def __post_init__(self) -> None:
                     f"{str(self.view.func.name)} and {str(self.view_inplace.func.name)} must either"
                     " both have CompositeImplicitAutograd kernels, or both not have composite kernels."
                 )
+        if self.view.has_composite_implicit_autograd_nested_tensor_kernel:
```
Comment (drisspg) [inline torchgen/model.py:2589]: > same comment as https://github.com/pytorch/pytorch/pull/82754/files#r950522386 (I don't think we need this)

**Comment by drisspg** [comment]
Closing cause this has been replaced with #84154
